pkg/metrics/jobmetrics.go

```diff
@@ -18,36 +18,45 @@ var (
 // KubeJobCollector is a prometheus collector that generates job sourced metrics.
 type KubeJobCollector struct {
 	KubeClusterCache clustercache.ClusterCache
+	metricsConfig    MetricsConfig
 }
 
 // Describe sends the super-set of all possible descriptors of metrics
 // collected by this Collector.
 func (kjc KubeJobCollector) Describe(ch chan<- *prometheus.Desc) {
-	ch <- prometheus.NewDesc("kube_job_status_failed", "The number of pods which reached Phase Failed and the reason for failure.", []string{}, nil)
+	disabledMetrics := kjc.metricsConfig.GetDisabledMetricsMap()
+
+	if _, ok := disabledMetrics["kube_pod_annotations"]; !ok {
+		ch <- prometheus.NewDesc("kube_job_status_failed", "The number of pods which reached Phase Failed and the reason for failure.", []string{}, nil)
+	}
 }
 
 // Collect is called by the Prometheus registry when collecting metrics.
 func (kjc KubeJobCollector) Collect(ch chan<- prometheus.Metric) {
-	jobs := kjc.KubeClusterCache.GetAllJobs()
-	for _, job := range jobs {
-		jobName := job.GetName()
-		jobNS := job.GetNamespace()
-
-		if job.Status.Failed == 0 {
-			ch <- newKubeJobStatusFailedMetric(jobName, jobNS, "kube_job_status_failed", "", 0)
-		} else {
-			for _, condition := range job.Status.Conditions {
-				if condition.Type == batchv1.JobFailed {
-					reasonKnown := false
-					for _, reason := range jobFailureReasons {
-						reasonKnown = reasonKnown || failureReason(&condition, reason)
-
-						ch <- newKubeJobStatusFailedMetric(jobName, jobNS, "kube_job_status_failed", reason, boolFloat64(failureReason(&condition, reason)))
-					}
+	disabledMetrics := kjc.metricsConfig.GetDisabledMetricsMap()
+
+	if _, ok := disabledMetrics["kube_pod_annotations"]; !ok {
+		jobs := kjc.KubeClusterCache.GetAllJobs()
```

Comment:
As a general point of Go style, aimed at avoiding excessive indentation, its recommended to do an early return, like so:

```suggestion

	if _, ok := disabledMetrics["kube_pod_annotations"]; ok {
	    return
	}
	jobs := kjc.KubeClusterCache.GetAllJobs()
```

Reply:
It has the added benefit of making the git diffs a little easier to parse, too.

Reply:
@michaelmdresser It can also reduce branches as well, which can make code way more performant. I don't know if it applies in this particular case, but it's a good reason to apply early returns/continues up and down the stack

Reply:
I've implemented this where I could. I know it's not super clean, but it's just the price we have to pay for this feature I guess.